### PR TITLE
[clang-tidy][NFC] optimize unused using decls performance

### PR DESCRIPTION
**Pull request by HerrCai0907**
Improve performance by moving the check forward to the matching stage

### CODE REVIEW

File: clang-tools-extra/clang-tidy/misc/UnusedUsingDeclsCheck.cpp [hunk 1]
```diff
@@ -25,6 +25,13 @@ AST_MATCHER_P(DeducedTemplateSpecializationType, refsToTemplatedDecl,
   return false;
 }
 
+AST_MATCHER_P(Type, getTagDecl, clang::ast_matchers::internal::Matcher<TagDecl>,
+              DeclMatcher) {
+  if (const auto *ND = Node.getAsTagDecl())
```
Comment (EugeneZelenko) [inline clang-tools-extra/clang-tidy/misc/UnusedUsingDeclsCheck.cpp:30]: Please do not use `auto` - type is not spelled explicitly.

Comment (HerrCai0907) [inline clang-tools-extra/clang-tidy/misc/UnusedUsingDeclsCheck.cpp:30]:
```suggestion
  if (const TagDecl *ND = Node.getAsTagDecl())
```

File: clang-tools-extra/clang-tidy/misc/UnusedUsingDeclsCheck.cpp [hunk 2]
```diff
@@ -25,6 +25,13 @@ AST_MATCHER_P(DeducedTemplateSpecializationType, refsToTemplatedDecl,
   return false;
 }
 
+AST_MATCHER_P(Type, getTagDecl, clang::ast_matchers::internal::Matcher<TagDecl>,
```
Comment (nicovank) [inline clang-tools-extra/clang-tidy/misc/UnusedUsingDeclsCheck.cpp:28]: `asTagDecl` maybe?

**Review by nicovank — APPROVED** [review]

**Comment by firewave** [comment]
This might help with #72300.

**Comment by HerrCai0907** [comment]
> This might help with #72300.

Yes. but not much. It can optimize about 5% execution time for some files in llvm projects.

**Pull request merged**